README.md
<div style="text-align:center;">

	<img src="http://smart-origin.com/images/so_logo_or_seul.png" height="73px">

</div>


# OSM Buildings for ArcGIS

This project is a fork of OSM Buildings JavaScript library for visualizing OpenStreetMap buildings into ArcGIS Api for Javascript 3.x.

The library is composed of three main components : 

- OSMBuildings original api from [kekscom](https://github.com/kekscom/osmbuildings)
- An interface to this api (ie adapter) named [ExternalInterface.js](https://github.com/smartorigin/osmbuildings/tree/so/arcgis/src/engines/ExternalInterface.js)
- A custom ArcGIS for Javascript Layer named [OSMBuildingsLayerForArcGIS](https://github.com/smartorigin/osmbuildings/tree/so/arcgis/src/engines/OSMBuildingsLayerForArcGIS.js)

The build process have been changed to support AMD loader, and especially dojo loader

Feel free to contact us for details and informations (usage, bugs, etc) at contact@smart-origin.com or via github.


# About OSM Buildings

<div style="text-align:center;">
	<img src="http://osmbuildings.org/logo.svg" width="100" height="88">
</div>


OSM Buildings is a JavaScript library for visualizing OpenStreetMap building geometry on interactive maps.

Example: http://osmbuildings.org/

The new WebGL version is located here: https://github.com/OSMBuildings/OSMBuildings
All variants of OSM Buildings will eventually land over there.

There is also documentation of OSM Buildings Server side:
https://github.com/kekscom/osmbuildings/blob/master/docs/server.md

## Deprecated methods and their replacements

  - `loadData()` => `load()`
  - `setData()` => `data()`
  - `setStyle()` => `style()`
  - `setDate()` => `date()`
  - `screenshot()` => no replacement
  - `getDetails()` => no replacement

**Example** http://osmbuildings.org/

It's safe use the [master branch](https://github.com/kekscom/osmbuildings/tree/master/dist/) for production.

For further information visit http://osmbuildings.org, follow [@osmbuildings](https://twitter.com/osmbuildings/) on Twitter or report issues [here on Github](https://github.com/kekscom/osmbuildings/issues/).


# Documentation

## Integration with ArcGIS

Declare a new package in your dojoConfig package information

	{name:'osmb-so', location:'osmbuildings-smartorigin/dist'}
  
Create a new layer and add it to the map
  
    define(["esri/map","osmb-so/OSMBuildingsLayerForArcGIS", "dojo/domReady!"], function (Map, OSMBuildingsLayerForArcGIS) {
      var map = new Map("map-area", {
        basemap: "streets",
        center: [5.72909, 45.18968],
        zoom:20
      });
      
      map.on('load', function() {
        try
        {
          map.addLayer(new OSMBuildingsLayerForArcGIS());
        }
        catch(error)
        {
          console.error('Unable to add OSMBuildings layer', error);
        }
    
      });
    });


## Integration with Leaflet

Link Leaflet and OSM Buildings files in your HTML head section.

~~~ html
<head>
  <link rel="stylesheet" href="http://cdn.leafletjs.com/leaflet-0.7.3/leaflet.css">
  <script src="http://cdn.leafletjs.com/leaflet-0.7.3/leaflet.js"></script>
  <script src="OSMBuildings-Leaflet.js"></script>
</head>
~~~

Initialize the map engine and add a map tile layer.<br>
Position is set to Berlin at zoom level 17, I'm using MapBox tiles here.

~~~ javascript
var map = new L.Map('map').setView([52.52020, 13.37570], 17);
new L.TileLayer('http://{s}.tiles.mapbox.com/v3/<YOUR KEY HERE>/{z}/{x}/{y}.png',
  { attribution: 'Map tiles &copy; <a href="http://mapbox.com">MapBox</a>', maxZoom: 17 }).addTo(map);
~~~

Add the buildings layer.

~~~ javascript
new OSMBuildings(map).load();
~~~

As a popular alternative, you could pass a <a href="http://www.geojson.org/geojson-spec.html">GeoJSON</a> FeatureCollection object.<br>
Geometry types Polygon, Multipolygon and GeometryCollection are supported.<br>
Make sure the building coordinates are projected in <a href="http://spatialreference.org/ref/epsg/4326/">EPSG:4326</a>.<br>
Height units m, ft, yd, mi are accepted, no given unit defaults to meters.

~~~ javascript
var geoJSON = {
  "type": "FeatureCollection",
  "features": [{
    "type": "Feature",
    "id": 134,
    "geometry": {
      "type": "Polygon",
      "coordinates": [[
        [13.37356, 52.52064],
        [13.37350, 52.51971],
        [13.37664, 52.51973],
        [13.37594, 52.52062],
        [13.37356, 52.52064]
      ]]
    },
    "properties": {
      "wallColor": "rgb(255,0,0)",
      "roofColor": "rgb(255,128,0)",
      "height": 500,
      "minHeight": 0
    }
  }]
};

new OSMBuildings(map).set(geoJSON);
~~~


## Integration with OpenLayers

Link OpenLayers and OSM Buildings files in your HTML head section.

~~~ html
<head>
  <script src="http://www.openlayers.org/api/OpenLayers.js"></script>
  <script src="OSMBuildings-OpenLayers.js"></script>
</head>
~~~

Initialize the map engine and add a map tile layer.<br>
Position is set to Berlin at zoom level 17.

~~~ javascript
var map = new OpenLayers.Map('map');
map.addControl(new OpenLayers.Control.LayerSwitcher());

var osm = new OpenLayers.Layer.OSM();
map.addLayer(osm);

map.setCenter(
  new OpenLayers.LonLat(13.37570, 52.52020)
    .transform(
      new OpenLayers.Projection('EPSG:4326'),
      map.getProjectionObject()
    ),
  17
);
~~~

Add the buildings layer.

~~~ javascript
new OSMBuildings(map).load();
~~~


# API

## Constructors

<table>
<tr>
<th>Constructor</th>
<th>Description</th>
</tr>

<tr>
<td>new OSMBuildings(map)</td>
<td>Initializes the buildings layer for a given map engine.<br>
Currently Leaflet and OpenLayers are supported.</td>
</tr>
</table>

Constants

<table>
<tr>
<th>Option</th>
<th>Type</th>
<th>Description</th>
</tr>

<tr>
<td>ATTRIBUTION</td>
<td>String</td>
<td>Holds OSM Buildings copyright information.</td>
</tr>

<tr>
<td>VERSION</td>
<td>String</td>
<td>Holds current version information.</td>
</tr>
</table>

Methods

<table>
<tr>
<th>Method</th>
<th>Description</th>
</tr>

<tr>
<td>style({Object})</td>
<td>Set default styles. See below for details.</td>
</tr>

<tr>
<td>date(new Date(2015, 15, 1, 10, 30)))</td>
<td>Set date/time for shadow projection.</td>
</tr>

<tr>
<td>each({Function})</td>
<td>A callback wrapper to override each feature's properties on read. Return false in order to skip a particular feature.<br>
Callback receives a feature object as argument.</td>
</tr>

<tr>
<td>click({Function})</td>
<td>A callback wrapper to handle click events on features.<br>
Callback receives an object { featureId{number,string}, lat{float}, lon{float} } as argument.</td>
</tr>

<tr>
<td>set({GeoJSON FeatureCollection})</td>
<td>Just add GeoJSON data to your map.</td>
</tr>

<tr>
<td>load({Provider})</td>
<td>Without parameter, it loads OpenStreetMap data tiles via an OSM Buildings proxy. This proxy enables transparent data filtering and caching.
Interface of such provider is to be published.</td>
</tr>
</table>

Styles

<table>
<tr>
<th>Option</th>
<th>Type</th>
<th>Description</th>
</tr>

<tr>
<td>color/wallColor</td>
<td>String</td>
<td>Defines the objects default primary color. I.e. #ffcc00, rgb(255,200,200), rgba(255,200,200,0.9)</td>
</tr>

<tr>
<td>roofColor</td>
<td>String</td>
<td>Defines the objects default roof color.</td>
</tr>

<tr>
<td>shadows</td>
<td>Boolean</td>
<td>Enables or disables shadow rendering, default: enabled</td>
</tr>
</table>


# Data

## OSM Tags used

<table>
<tr>
<th>GeoJSON property</th>
<th>OSM Tags</th>
</tr>

<tr>
<td>height</td>
<td>height, building:height, levels, building:levels</td>
</tr>

<tr>
<td>minHeight</td>
<td>min_height, building:min_height, min_level, building:min_level</td>
</tr>

<tr>
<td>color/wallColor</td>
<td>building:color, building:colour</td>
</tr>

<tr>
<td>material</td>
<td>building:material, building:facade:material, building:cladding</td>
</tr>

<tr>
<td>roofColor</td>
<td>roof:color, roof:colour, building:roof:color, building:roof:colour</td>
</tr>

<tr>
<td>roofMaterial</td>
<td>roof:material, building:roof:material</td>
</tr>

<tr>
<td>shape</td>
<td>building:shape[=cylinder,sphere]</td>
</tr>

<tr>
<td>roofShape</td>
<td>roof:shape[=dome]</td>
</tr>

<tr>
<td>roofHeight</td>
<td>roof:height</td>
</tr>
</table>


# Developpers

## Configure your environment

    npm install
  
## Build for your environment
  
Edit config.js and choose which adapter you want to build
  
    exports.engines = ['Leaflet', 'OpenLayers', 'OL3', 'ExternalInterface'];
  
For building ArcGIS adapter, please edit config.js and choose prefix_amd and suffix_amd as replacement of prefix and suffix in library

  
Goto build folder :  
  
    sh build-debug.sh #for debug version only
    sh build.sh



CHANGELOG.md
# Changelog

## 0.2.3b @ 2015-12-16

**Features**

- Add new Adapter ExternalInterface
- Add an ArcGIS for javascript 3.x layer implementation (OSMBuildingsLayerForArcGIS)

**Fixes**

- build remove debugger statements before hinting
- use new Color function instead of Color.parse

## 0.2.2b @ 2014-11-25

**Features**

- detection for circular objects

**Fixes**

- several fixes for radius handling
- several click-hit issues fixed


## 0.2.1b @ 2014-09-24

**Features**

- pyramidal buildungs and roofs
- buildings parts are a logical unit now (i.e. OSM relations)
- explicitly injecting OSMBuildings class into global namespace [#66](https://github.com/kekscom/osmbuildings/issues/66)


## 0.2.0b @ 2014-09-04

*From this version on, OSM Buildings is entering beta phase \o/*

**Features**

- Buildings are clickable now, use .click(function(featureId) {...})
- Massive improvements in GeoJSON reading, bigger set of properties and GeometryCollections are supported
- Ambient shadows for buildings added
- Introduced a data service for filtering and caching OSM data, results in massive speedup for loading
- Geometry: cones enabled, also used as an interim replacement for domes
- Tested device accelerated perspective aka Amazon's 'Dynamic Perspective', but disabled again in favor for performance
- Successfully tested with LeafletJS 0.8 and OpenLayers 2.13.1
- Code size reduced from 10.23 to 9.44k (all gzipped)

**Fixes**

- Height scale fixed
- Relation properties precedence fixed
- Simple buildings layer refactored
- Flipped perspective on some latitudes fixed
- Tried requestAnimationFrame, but needed to drop again for IE and iOS


## 0.1.9a @ 2013-10-17

- multipolygon support added
- backend removed, now using web services with GeoJSON or OSM Overpass XAPI
- vector data is subdivided into tiles
- data tiles are cached
- fix for chained method calls
- fix for flat buildings from rendering tall buildings too
- min zoom level decreased to 15
- fix for setStyle() removing shadows
- material color mapping added
- HSLA color support added
- support for W3C named colors added
- CORS-XHR support for MSIE added
- cylindric object rendering added
- API is now documented in GitHub README
- map engine adapters simplified
- minHeight and height units for GeoJSON enabled
- very simple fix for building occlusion
- successful tests with LeafletJS 0.6.4


## 0.1.8a @ 2013-03-10

- on layer removal from map, OSM Buildings is not destroyed anymore
- introduced multiple rendering layers
- improved simplification algorithm, inspired by Vladimir Agafonkin (http://mourner.github.com/simplify-js)
- initial version of objects draw order (farthest first, lower first)
- directional wall shading added
- building shadows added
- shadow date / time dependency added, inspired by Vladimir Agafonkin /https://github.com/mourner/suncalc)
- `min_height` support added (requires backend change)
- color / style table handling improved
- rendering tests added
- successful tests with LeafletJS 0.5.1
- recommendation: reduce building `$heightScale` in backend server config down to 1.2


## 0.1.7a @ 2012-10-10

- adding OpenLayers support, credits to Jérémy Judéaux (https://github.com/Volune)
- aligning Layer naming convention to engines
- fixing some rare cases where layer got removed


## 0.1.6a @ 2012-09-04

- GeoJSON: min zoom removed
- GeoJSON: height property re-enabled
- GeoJSON: multi polygons enabled
- Examples are rebuilt entirely
- Roof colors are re-enbled
- JSHint is now part of the build process


## 0.1.5a

- support for GeoJSON improved
- deep integration with Leaflet in order to avoid jittery movement
- enabled individual building colors
- polygon winding fixed


## 0.1.0a

- GeoJSON support added
- method chaining added
- adding converter PostGIS > MySQL
- data for Frankfurt added
- made either MySQL or PostGIS (Mapnik) fully configurable
- lat/lon order of your coordinates is configurable
- polygon direction is forced to be clockwise
- simpler initialization process


src/Data.js

var Data = {

  loadedItems: {}, // maintain a list of cached items in order to avoid duplicates on tile borders
  items: [],

  getPixelFootprint: function(buffer) {
    var footprint = new Int32Array(buffer.length),
      px;

    for (var i = 0, il = buffer.length-1; i < il; i+=2) {
      px = geoToPixel(buffer[i], buffer[i+1]);
      footprint[i]   = px.x;
      footprint[i+1] = px.y;
    }

    footprint = simplifyPolygon(footprint);
    if (footprint.length < 8) { // 3 points & end==start (*2)
      return;
    }

    return footprint;
  },

  resetItems: function() {
    this.items = [];
    this.loadedItems = {};
    HitAreas.reset();
  },

  addRenderItems: function(data, allAreNew) {
    var item, scaledItem, id;
    var geojson = GeoJSON.read(data);
    for (var i = 0, il = geojson.length; i < il; i++) {
      item = geojson[i];
      id = item.id || [item.footprint[0], item.footprint[1], item.height, item.minHeight].join(',');
      if (!this.loadedItems[id]) {
        if ((scaledItem = this.scale(item))) {
          scaledItem.scale = allAreNew ? 0 : 1;
          this.items.push(scaledItem);
          this.loadedItems[id] = 1;
        }
      }
    }
    fadeIn();
  },

  scale: function(item) {
    var
      res = {},
      // TODO: calculate this on zoom change only
      zoomScale = 6 / pow(2, ZOOM-MIN_ZOOM); // TODO: consider using HEIGHT / (global.devicePixelRatio || 1)

    if (item.id) {
      res.id = item.id;
    }

    res.height = min(item.height/zoomScale, MAX_HEIGHT);

    res.minHeight = isNaN(item.minHeight) ? 0 : item.minHeight / zoomScale;
    if (res.minHeight > MAX_HEIGHT) {
      return;
    }

    res.footprint = this.getPixelFootprint(item.footprint);
    if (!res.footprint) {
      return;
    }
    res.center = getCenter(res.footprint);

    if (item.radius) {
      res.radius = item.radius*PIXEL_PER_DEG;
    }
    if (item.shape) {
      res.shape = item.shape;
    }
    if (item.roofShape) {
      res.roofShape = item.roofShape;
    }
    if ((res.roofShape === 'cone' || res.roofShape === 'dome') && !res.shape && isRotational(res.footprint)) {
      res.shape = 'cylinder';
    }

    if (item.holes) {
      res.holes = [];
      var innerFootprint;
      for (var i = 0, il = item.holes.length; i < il; i++) {
        // TODO: simplify
        if ((innerFootprint = this.getPixelFootprint(item.holes[i]))) {
          res.holes.push(innerFootprint);
        }
      }
    }

    var color;

    if (item.wallColor) {
      if ((color = new Color(item.wallColor))) {
        color = color.alpha(ZOOM_FACTOR);
        res.altColor  = ''+ color.lightness(0.8);
        res.wallColor = ''+ color;
      }
    }

    if (item.roofColor) {
      if ((color = new Color(item.roofColor))) {
        res.roofColor = ''+ color.alpha(ZOOM_FACTOR);
      }
    }

    if (item.relationId) {
      res.relationId = item.relationId;
    }
    res.hitColor = HitAreas.idToColor(item.relationId || item.id);

    res.roofHeight = isNaN(item.roofHeight) ? 0 : item.roofHeight/zoomScale;

    if (res.height+res.roofHeight <= res.minHeight) {
      return;
    }

    return res;
  },

  set: function(data) {
    this.isStatic = true;
    this.resetItems();
    this._staticData = data;
    this.addRenderItems(this._staticData, true);
  },

  load: function(src, key) {
    this.src = src || DATA_SRC.replace('{k}', (key || 'anonymous'));
    this.update();
  },

  update: function() {
    this.resetItems();

    if (ZOOM < MIN_ZOOM) {
      return;
    }

    if (this.isStatic && this._staticData) {
      this.addRenderItems(this._staticData);
      return;
    }

    if (!this.src) {
      return;
    }

    var
      tileZoom = 16,
      tileSize = MAP_TILE_SIZE,
      zoomedTileSize = ZOOM > tileZoom ? tileSize <<(ZOOM-tileZoom) : tileSize >>(tileZoom-ZOOM),
      minX = ORIGIN_X/zoomedTileSize <<0,
      minY = ORIGIN_Y/zoomedTileSize <<0,
      maxX = ceil((ORIGIN_X+WIDTH) /zoomedTileSize),
      maxY = ceil((ORIGIN_Y+HEIGHT)/zoomedTileSize),
      x, y;

    var scope = this;
    function callback(json) {
      scope.addRenderItems(json);
    }

    for (y = minY; y <= maxY; y++) {
      for (x = minX; x <= maxX; x++) {
        this.loadTile(x, y, tileZoom, callback);
      }
    }
  },
  
  loadTile: function(x, y, zoom, callback) {
    var s = 'abcd'[(x+y) % 4];
    var url = this.src.replace('{s}', s).replace('{x}', x).replace('{y}', y).replace('{z}', zoom);
    return Request.loadJSON(url, callback);
  }
};


src/engines/OSMBuildingsLayerForArcGIS.js
/**
 * OSMBuildingsLayerForArcGIS dojo module
 * Created on 16/12/2015.
 * @author Tony Georges - Smart/Origin
 * @copyright (c)2015, Smart Origin SARL
 * @license
 * Copyright (c) 2015 Smart Origin SARL
 *
 * Redistribution and use in source and binary forms, with or without modification, are
 * permitted provided that the following conditions are met:
 *
 *  1. Redistributions of source code must retain the above copyright notice, this list of
 *  conditions and the following disclaimer.

 *  2. Redistributions in binary form must reproduce the above copyright notice, this list
 *  of conditions and the following disclaimer in the documentation and/or other materials
 *  provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
 * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
 * TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
define(['dojo/_base/declare', "dojo/_base/lang", 'dojo/_base/array', 'dojo/dom-construct', 'dojo/dom-style',
        'esri/request', 'esri/config', 'esri/tasks/QueryTask', 'dojo/has!esri-3x?esri/tasks/query:esri/tasks/support/Query', 'dojo/has!esri-3x?esri/layers/layer:esri/layers/Layer',
        'esri/geometry/Point', './OSMBuildings-ExternalInterface'],
       function (declare, lang, array, domConstruct, domStyle,
                 esriRequest, esriConfig, QueryTask, Query, Layer, Point, ExternalInterface)
       {

         /**
          * ArcGIS for JavaScript OSMBuildings custom layer.
          * Encapsulate access to OSMBuildings "canvas" implementation through a simple ExternalInterface version of the OSMBuildings canvas api.
          *
          * This layer was originally provided in OSMBuildings engines, but was not compatible with dojo AMD loader.
          * All the capabilities provided in first version were kept in place and will be fully tested and validated soon (especially the connection with a backend FeatureLayer for attributes)
          *
          * Feel free to contact me for details and informations (usage, bugs, etc)
          *
          * @class so.arcgis.layers.BuildingsLayer
          * @extends esri.layer.Layer
          */
         var BuildingsLayer           = declare('so.arcgis.layers.BuildingsLayer', [Layer], {/** @lends so.arcgis.layers.BuildingsLayer.prototype */
           _externalInterface           : null,
           _container: null,
           _tileInfo : null,
           _mode     : 0,// ON_DEMAND|SNAPSHOT
           _heightAttribute: '',
           _oidField       : null,  // will be overridden from meta query
           _query          : null,
           _task           : null,
           _oids           : null, //track current objectids to mark new/old for animation fade in
           _featureExt     : null,//current feature extent, 1.5 map extent
           _suspendOnPan   : false,//whether to suspend drawing during map panning. default is false.
           // set to true if performance is not optimal (non-Chrome browsers);

           /**
            * @name BuildingsLayerOptions
            * @class This is an object literal that specify the options for each BuildingsLayer.
            * @property {string} heightAttribute required. name of the attribute for height;
            * @property {number} [defaultHeight] optional. default Height to use if the height value is 0. default=0;
            * @property {number} [heightScaleRatio] optional. number used to multiple the value from service. default=1;
            * @property {number} [extentScaleRatio] optional. extra buffer on map extent to load features to reduce server traffic. default=1.5;
            * @property {esri.tasks.Query} [query] optional. query set on the feature layer for retrieving data.
            * @property {int} [mode] optional. so.arcgis.layers.BuildingsLayer.MODE_ONDEMAND | MODE_SNAPSHOT. default=ON_DEMAND
            * @property {Object} [style] object with wallColor, roofColor, altColor, shadowColor, shadowBlurColor (#ffcc00' 'rgb(255,200,200)' 'rgba(255,200,200,0.9))
            */
           /**
            * Create an BuildingsLayer
            * @name BuildingsLayer
            * @constructor
            * @class This class is a BuildingsLayer, such as a polygon feature layer with a height attribute.
            * @param {string||FeatureCollection} url
            * @param {BuildingsLayerOptions} opts
            */
           constructor: function (url, opts)
           {
             // Manually call superclass constructor with required arguments
             if (!(!!document.createElement('canvas').getContext))
             {
               throw new Error('Canvas unsupported. Try different browser');
             }
             this.inherited(arguments);
             opts                   = opts || {};
             this._heightAttribute = opts.heightAttribute;
             this._mode            = opts.mode || BuildingsLayer.MODE_ONDEMAND;
             this._heightScaleRatio = opts.heightScaleRatio || 1;
             this._extentScaleRatio = opts.extentScaleRatio || 1.5;
             this._defaultHeight    = opts.defaultHeight || 0;

             //Create external interface
             this._externalInterface = new ExternalInterface();

             /* Apply MIN ZOOM */
             this._externalInterface.setMinZoom(18);
             /* Apply default style */
             this._applyStyle(opts.style);

             // Deal with feature collection
             if (url)
             {
               if (lang.isObject(url) && url.featureSet)
               {
                 this._mode = BuildingsLayer.MODE_SNAPSHOT;
                 this._setFeatures(url.featureSet.features);
                 this.loaded = true;
                 this.onLoad(this);
               }
               else
               {
                 this._url = url;
                 // get meta data for layer
                 new esriRequest({
                   url              : this._url,
                   content: {
                     f: 'json'
                   },
                   callbackParamName: 'callback'
                 }).then(lang.hitch(this, this._initLayer));
                 this._query = new Query();
                 lang.mixin(this._query, opts.query);
                 lang.mixin(this._query, {
                   returnGeometry     : true,
                   outSpatialReference: {wkid: 4326}
                 });
                 this._task = new QueryTask(url);
                 this._task.on('complete', lang.hitch(this, this._onQueryComplete));
                 this._task.on('error', esriConfig.defaults.io.errorHandler);
               }
             }
             else
             {
               this._mode = BuildingsLayer.MODE_SNAPSHOT;
               this._setFeatures([]);
               this.loaded = true;
               this.onLoad(this);
             }
           },

           /**********************
            * @see http://help.arcgis.com/en/webapi/javascript/arcgis/samples/exp_rasterlayer/javascript/RasterLayer.js
            * Internal Properties
            *
            * _map
            * _element
            * _context
            * _mapWidth
            * _mapHeight
            * _connects
            **********************/
           _setMap         : function (map, container, ind, lod)
           {
             this._map = map;
             this.inherited(arguments);

             var element = domConstruct.create('div', {

               style: 'position: absolute; left: 0px; top: 0px;width : ' + map.width + 'px;height: ' + map.height + 'px'
             }, container);

             // allow attribution widget to add copyright text
             this.suspended = false;
             /* Attribution */
             this.copyright = this._externalInterface.getAttribution() + (this.copyright?(',' + this.copyrightText):'');

             this._element = element;

             /* Append OSMBuildings surface to layer div*/
             this._container = this._externalInterface.appendTo(element);

             /* TODO : setSize */
             this._externalInterface.setSize(map);
             // calculate origins
             if (map.layerIds.length == 0 || !map.getLayer(map.layerIds[0]).tileInfo)
             {
               throw new Error('must have at least one tiled layer added before this layer');
             }
             this._tileInfo = map.getLayer(map.layerIds[0]).tileInfo;
             this._externalInterface.setTileSize(this._tileInfo.width);
             //TODO : setZOOM
             this._externalInterface.setZoom(map.getLevel());
             this._setOrigin();
             // TODO : LOAD
             this._externalInterface.load();
             this._loadData();
             // Event connections
             this._connects = [];
             this._connects.push(map.on('resize', lang.hitch(this, this._onResize)));
             this._connects.push(map.on('pan', lang.hitch(this, this._onPan)));
             this._connects.push(map.on('extent-change', lang.hitch(this, this._onExtentChange)));
             this._connects.push(map.on('zoom-start', lang.hitch(this, this._onZoomStart)));
             return element;
           },
           // esri.layers.Layer.method
           _unsetMap: function (map, container)
           {
             if (this._externalInterface)
             {
               this._container.parentNode.removeChild(this._container);
               this._externalInterface = null;
             }
             array.forEach(this._connects, function (handle)
             {
               handle.remove();
             });
             if (this._element)
             {
               container.removeChild(this._element);
             }
             this._map     = null;
             this._element = null;
           },
           _initLayer: function (json)
           {
             //dojo.mixin(this, json);
             this.setMinScale(json.minScale || 0);
             this.setMaxScale (json.maxScale || 0);
             this.copyrightText = json.copyrightText;
             array.some(json.fields, function (field, i)
             {
               if (field.type == 'esriFieldTypeOID')
               {
                 this._oidField = field.name;
                 return true;
               }
               return false;
             }, this);
             this._query.outFields = [this._oidField, this._heightAttribute];
             this.loaded           = true;
             this.onLoad(this);
           },
           _setOrigin: function (dx, dy)
           {
             var resolution = this._tileInfo.lods[this._map.getLevel()].resolution;
             var topLeft = this._map.toMap(new Point(0, 0));
             var x       = Math.round((topLeft.x - this._tileInfo.origin.x) / resolution);
             var y       = Math.round((this._tileInfo.origin.y - topLeft.y) / resolution);
             //setOrigin
             this._externalInterface.setOrigin({x: x + (dx || 0), y: y + (dy || 0)});
             //setSize
             this._externalInterface.setSize({width: this._map.width, height: this._map.height});

           },
           _onResize : function (extent, width, height)
           {
             if (this._externalInterface)
             {
               //setSize
               this._externalInterface.setSize({width: width, height: height});
               //render
               this._externalInterface.render();
             }
           },
           _onPan    : function (event)//)extent, delta)
           {
             if (this._suspendOnPan)
             {
               domStyle.set(this._container, {
                 left: event.delta.x + 'px',
                 top : event.delta.y + 'px'
               });
               //moveCam(-delta.x, -delta.y);
             }
             else
             {
               this._setOrigin(-event.delta.x, -event.delta.y);
               //render
               this._externalInterface.render();
             }
           },

           _onExtentChange: function (event)
           {
             domStyle.set(this._container, {
               left: 0,
               top : 0
             });
             this._setOrigin();
             //moveCam
             this._externalInterface.moveCam({x: 0, y: 0});
             if (event.levelChange)
             {
               //Execute onZoomEnd
               this._externalInterface.onZoomEnd({
                           zoom: this._map.getLevel()
                         });

               if (this.isVisibleAtScale(this._map.getScale()))
               {
                 this._loadData();
               }
               else
               {
                 // clear canvas.
                 this._externalInterface.clear();
               }
             }
             else
             {
               //Call onMoveEnd
               this._externalInterface.onMoveEnd();
               if (this._featureExt && !this._featureExt.contains(extent))
               {
                 this._loadData();
               }
             }
           },
           _onZoomStart   : function (extent, zoomFactor, anchor, level)
           {
             //CallonZoomStart
             this._externalInterface.onZoomStart();
           },
           _setFeatures   : function (features)
           {
             var oids   = {};
             var jfs  = [];
             this._oids = this._oids || {};
             for (var i = 0; i < features.length; i++)
             {
               var f = features[i];

               var oid = f.attributes[this._oidField];
               var gj  = {
                 type      : 'Feature',
                 geometry: {
                   type       : 'Polygon',
                   coordinates: f.geometry.rings
                 },
                 properties: {
                   height: (f.attributes[this._heightAttribute] || this._defaultHeight) * this._heightScaleRatio,
                   isNew : !this._oids[oid]
                 }
               };
               // find out the y coords range for sorting
               var minY = maxY = f.geometry.rings[0][0][1];
               for (var j = 0; j < f.geometry.rings.length; j++)
               {
                 for (var k = 0; k < f.geometry.rings[j].length; k++)
                 {
                   minY = Math.min(minY, f.geometry.rings[j][k][1]);
                   maxY = Math.max(maxY, f.geometry.rings[j][k][1]);
                 }
               }
               gj.minY   = minY;
               gj.maxY = maxY;
               jfs[i]  = gj;
               oids[oid] = f;
             }
             // sort features by height and y coord desc for potential drawing improvement
             jfs.sort(function (a, b)
                      {
                        // if polygon a is completely north of b then put a first.
                        // otherwise put the taller one first.
                        // this ensures north/taller poly draw first
                        if (a.maxY < b.minY)
                        {
                          return 1;
                        }
                        else if (a.minY > b.maxY)
                        {
                          return -1;
                        }
                        else
                        {
                          return b.properties.height - a.properties.height
                        }
                      });
             this._oids = oids;

             //TODO : test
             this._externalInterface.addRenderItems({
                                                      type: 'FeatureCollection',
                                                      features: jfs
                                                    });

             if (this._style)
             {
               //style
               this._externalInterface.style(this._style);
             }
           },
           _loadData      : function ()
           {
             if (this._mode == BuildingsLayer.MODE_SNAPSHOT)
             {
               if (this._oids)
               {
                 return;
               }
               else
               {
                 this._query.geometry = null;
                 this._query.where    = this._query.where || '1=1';

               }
             }
             else
             {
               this._featureExt     = this._map.extent.expand(this._extentScaleRatio);
               this._query.geometry = this._featureExt;
             }
             this._task.execute(this._query);


           },
           _onQueryComplete: function (featureSet)
           {
             this._setFeatures(featureSet.features);
           },

           _applyStyle : function(style)
           {
             style = style || {}
             var wallColor = style.wallColor || 'rgba(219, 217, 213)',
                 altColor  = style.altColor || 'rgba(194, 192, 188)',
                 roofColor = style.roofColor || 'rgba(252, 252, 252,1)',
                 shadowColor = style.shadowColor || '#444444',
                 shadowBlurColor = style.shadowBlurColor || '#111111';
             this._externalInterface.setStyle(wallColor, altColor, roofColor, shadowColor, shadowBlurColor);

           }
         });
         BuildingsLayer.MODE_ONDEMAND = 0;
         BuildingsLayer.MODE_SNAPSHOT = 1;

         return BuildingsLayer;
       });